src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/IndexBuilder.java

```diff
@@ -300,7 +300,14 @@ private boolean shouldIndex(Element element) {
     }
 
     private static Character keyCharacter(String s) {
-        return s.isEmpty() ? '*' : Character.toUpperCase(s.charAt(0));
+        // Use first valid java identifier start character as key,
+        // or '*' for strings that do not contain one.
+        for (int i = 0; i < s.length(); i++) {
+            if (Character.isJavaIdentifierStart(s.charAt(i))) {
+                return Character.toUpperCase(s.charAt(i));
```

Comment:
maybe use the overload that takes a locale, and use `Locale.ROOT`

Reply:
I just learned that `Character.toUpperCase` does not have a locale-sensitive overload, we'd have to use `String.toUpperCase` for that. I don't think it's that important in this case, so I'm leaving the code as it is.